src/persist-client/src/lib.rs

```diff
@@ -1453,6 +1453,48 @@ mod tests {
         assert_eq!(snap.read_all().await, all_ok(&data[..], 3));
     }
 
+    // Regression test for #13802. A snapshot read would error if its read handle tried
+    // to heartbeat without having recent enough consensus state to contain its reader id.
+    #[tokio::test]
+    async fn regression_snapshot_readers() {
+        mz_ore::test::init_logging();
+
+        let data = vec![
+            (("1".to_owned(), "one".to_owned()), 1, 1),
+            (("2".to_owned(), "two".to_owned()), 2, 1),
+            (("3".to_owned(), "three".to_owned()), 3, 1),
+        ];
+
+        let (mut write, mut read) = new_test_client()
+            .await
+            .expect_open::<String, String, u64, i64>(ShardId::new())
+            .await;
+        write
+            .expect_append(&data[..2], write.upper().clone(), vec![3])
+            .await;
+
+        // initialize our snapshot's read handle with a heartbeat timestamp of 0
+        read.cfg.now = NowFn::from(|| 0);
+        let mut snap = read.expect_snapshot(1).await;
+        assert_eq!(snap.handle.last_heartbeat, 0);
+
+        // Fast-forward time enough so the following snapshot read will be forced to heartbeat
+        let near_future = NowFn::from(|| {
+            PersistConfig::FAKE_READ_LEASE_DURATION
```

Comment:
if you do the above comment, this could be something like `snap.handle.last_heartbeat + FAKE_READ_LEASE_DURATION` (though you'd have to compute the answer before the closure so it doesn't hold on to snap)